app/upgrades/v17/upgrades.go

```diff
@@ -1,105 +1,70 @@
 	mm *module.Manager,
 	configurator module.Configurator,
 	bpm upgrades.BaseAppParamManager,
 	keepers *keepers.AppKeepers,
 ) upgradetypes.UpgradeHandler {
 	return func(ctx sdk.Context, plan upgradetypes.Plan, fromVM module.VersionMap) (module.VersionMap, error) {
+		poolLinks := []gammmigration.BalancerToConcentratedPoolLink{}
+		fullRangeCoinsUsed := sdk.Coins{}
+
 		// Run migrations before applying any other state changes.
 		// NOTE: DO NOT PUT ANY STATE CHANGES BEFORE RunMigrations().
 		migrations, err := mm.RunMigrations(ctx, configurator, fromVM)
 		if err != nil {
 			return nil, err
 		}
 
-		// Get community pool address.
 		communityPoolAddress := keepers.AccountKeeper.GetModuleAddress(distrtypes.ModuleName)
 
-		// fullRangeCoinsUsed tracks the coins we use in the below for loop from the community pool to create the full range position for each new pool.
-		fullRangeCoinsUsed := sdk.NewCoins()
-
-		poolLinks := []gammmigration.BalancerToConcentratedPoolLink{}
-
-		assetPairs := InitializeAssetPairs(ctx, keepers)
-
-		for _, assetPair := range assetPairs {
-			// Create a concentrated liquidity pool for asset pair.
-			clPool, err := keepers.GAMMKeeper.CreateConcentratedPoolFromCFMM(ctx, assetPair.LinkedClassicPool, assetPair.BaseAsset, assetPair.SpreadFactor, TickSpacing)
-			if err != nil {
-				return nil, err
-			}
-			clPoolId := clPool.GetId()
-			clPoolDenom := cltypes.GetConcentratedLockupDenomFromPoolId(clPoolId)
-
-			// Add the pool link to the list of pool links (we set them all at once later)
-			poolLinks = append(poolLinks, gammmigration.BalancerToConcentratedPoolLink{
-				BalancerPoolId: assetPair.LinkedClassicPool,
-				ClPoolId:       clPoolId,
-			})
-
-			// Determine the amount of baseAsset that can be bought with 1 OSMO.
-			oneOsmo := sdk.NewCoin(QuoteAsset, sdk.NewInt(1000000))
-			linkedClassicPool, err := keepers.PoolManagerKeeper.GetPool(ctx, assetPair.LinkedClassicPool)
-			if err != nil {
-				return nil, err
-			}
-			respectiveBaseAsset, err := keepers.GAMMKeeper.CalcOutAmtGivenIn(ctx, linkedClassicPool, oneOsmo, assetPair.BaseAsset, sdk.ZeroDec())
-			if err != nil {
-				return nil, err
-			}
-
-			// Create a full range position via the community pool with the funds we calculated above.
-			fullRangeCoins := sdk.NewCoins(respectiveBaseAsset, oneOsmo)
-			_, actualBaseAmtUsed, actualQuoteAmtUsed, _, err := keepers.ConcentratedLiquidityKeeper.CreateFullRangePosition(ctx, clPoolId, communityPoolAddress, fullRangeCoins)
-			if err != nil {
-				return nil, err
-			}
-
-			// Track the coins used to create the full range position (we manually update the fee pool later all at once).
-			fullRangeCoinsUsed = fullRangeCoinsUsed.Add(sdk.NewCoins(sdk.NewCoin(QuoteAsset, actualQuoteAmtUsed), sdk.NewCoin(assetPair.BaseAsset, actualBaseAmtUsed))...)
-
-			// If pair was previously superfluid enabled, add the cl pool's full range denom as an authorized superfluid asset.
-			if assetPair.Superfluid {
```

Comment:
Let's remove Superfluid` field from the struct as it is not used anymore

Reply:
Superfluid is used in AssetPairsForTestsOnly.

Dev asked for a method (InitializeAssetPairs) that retrieves the current state of the pools provided instead of hard coding all the expected values.

I kept these hard coded values as AssetPairsForTestsOnly, which is essentially used to seed tests so they match what is on mainnet, and we can then test to see if InitializeAssetPairs initializes an assetList to an exact replica of what is saved on AssetPairsForTestsOnly.

Reply:
Why do we do the check [here](https://github.com/osmosis-labs/osmosis/blob/bedf1a5ee3a80a7f956f15cdaec09c3a99e39c0d/app/upgrades/v17/upgrades.go#L230) if it already present in the asset pair and can be given as parameter?

Reply:
This is a good point and allowed me to drastically simplify the upgrade handler here https://github.com/osmosis-labs/osmosis/pull/5959/commits/34cff152f94a197bd18b28f5a92d1d60f1f2a103

By sharing the AssetList abstraction, we don't need to have separate methods for testnet and mainnet, the only differing logic is how we select the AssetList in the first place!